src/libraries/System.Linq/src/System/Linq/Range.SpeedOpt.cs

```diff
@@ -16,23 +19,55 @@ public override IEnumerable<TResult> Select<TResult>(Func<int, TResult> selector
 
             public int[] ToArray()
             {
-                int[] array = new int[_end - _start];
-                Fill(array, _start);
+                int start = _start;
+                int[] array = new int[_end - start];
+                Fill(array, start);
                 return array;
             }
 
             public List<int> ToList()
             {
-                List<int> list = new List<int>(_end - _start);
-                Fill(SetCountAndGetSpan(list, _end - _start), _start);
+                int start = _start;
+                int end = _end;
+                List<int> list = new List<int>(end - start);
+                Fill(SetCountAndGetSpan(list, end - start), start);
                 return list;
             }
 
-            private static void Fill(Span<int> destination, int value)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static void Fill(Span<int> destination, int start)
             {
-                for (int i = 0; i < destination.Length; i++, value++)
+                // The improvements in the compiler now allow us to write the loop as Vector256.
+                // Manually unrolling to Vector128x2 for SSE2/NEON is no longer profitable.
```

Comment:
Can we just use `Vector<T>`? e.g.
```C#
static void Fill(Span<int> destination, int value)
{
    Debug.Assert(Vector<int>.Count * sizeof(int) * 8 <= 512);

    ref int c = ref MemoryMarshal.GetReference(destination);
    ref int end = ref Unsafe.Add(ref c, destination.Length);

    if (Vector.IsHardwareAccelerated && destination.Length >= Vector<int>.Count)
    {
        var init = new Vector<int>((ReadOnlySpan<int>)new int[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 });
        var increment = new Vector<int>(Vector<int>.Count);
        var current = new Vector<int>(value) + init;

        ref int oneVectorFromEnd = ref Unsafe.Subtract(ref end, Vector<int>.Count);
        do
        {
            current.StoreUnsafe(ref c);
            current += increment;
            c = ref Unsafe.Add(ref c, Vector<int>.Count);
        }
        while (!Unsafe.IsAddressLessThan(ref oneVectorFromEnd, ref c));

        value = current[0];
    }

    while (Unsafe.IsAddressLessThan(ref c, ref end))
    {
        c = value++;
        c = ref Unsafe.Add(ref c, 1);
    }
}
```

Reply:
It is definitely an option, but I initially opted out of it to have better performance on ARM64. As of now, there are pros and cons.

`Vector<T>` variant
Pros:
- Takes advantage of AVX512
- Does not rely on JIT/ILC being able to perfectly optimize V256 as V128x2 for AdvSimd/SSE2
- Consistent with how other vectorized code is written
- Better performance on Mono? (as @EgorBo mentioned, Mono regressions are an issue that needs to be tracked)

Cons:
- Does not help 0..15 i.e. `Enumerable.Range(0, 10).ToArray()` , which is fairly common, on AVX512
- Has reduced gains on wide ARM64 cores (M series can dispatch V128x2 loop iteration in a single cycle, ARM Cortex-X3/4 should be able to do the same). This also applies to default NativeAOT target which would use 128x2 with SSE2. I expect x86_64 behavior to be very similar.

Do you think `Vector<T>` is better for the time being?

Reply:
> Takes advantage of AVX512

Not by default.  By default, `Vector<T>` will be either 128-bits or 256-bits.

> Do you think Vector<T> is better for the time being?

Yes